micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -15,59 +15,118 @@
+    private static final String TIMER_NAME = "mongodb.driver.commands";
+    private static final String TIMER_DESC = "Timer of mongodb commands";
+    private static final int DEFAULT_COLLECTION_NAMES_CACHE_MAX_SIZE = 1000;
+    private static final int DEFAULT_LOG_EVERY_NTH_OPERATION_INTERVAL = 100;
 
     private final MeterRegistry registry;
+    private final MongoCommandUtil mongoCommandUtil;
+    private final ConcurrentHashMap<Integer, String> collectionNamesCache;
+    private final int collectionNamesCacheMaxSize;
+    @Nullable
+    private final EveryNthOperation collectionNamesCacheOverflowErrorTracker;
 
     public MongoMetricsCommandListener(MeterRegistry registry) {
+        this(registry, DEFAULT_COLLECTION_NAMES_CACHE_MAX_SIZE, DEFAULT_LOG_EVERY_NTH_OPERATION_INTERVAL);
+    }
+
+    public MongoMetricsCommandListener(MeterRegistry registry, int collectionNamesCacheMaxSize, int collectionNamesCacheOverflowErrorLogInterval) {
+        this(registry, collectionNamesCacheMaxSize, collectionNamesCacheOverflowErrorLogInterval < 1 ? null : new EveryNthOperation(collectionNamesCacheOverflowErrorLogInterval));
+    }
+
+    // VisibleForTesting
+    MongoMetricsCommandListener(MeterRegistry registry, int collectionNamesCacheMaxSize, @Nullable EveryNthOperation collectionNamesCacheOverflowErrorTracker) {
         this.registry = registry;
+        this.mongoCommandUtil = new MongoCommandUtil();
+        this.collectionNamesCache = new ConcurrentHashMap<>();
+        if (collectionNamesCacheMaxSize < 1) {
+            throw new IllegalArgumentException("collectionNamesCacheMaxSize must be set to a positive value");
+        }
+        this.collectionNamesCacheMaxSize = collectionNamesCacheMaxSize;
+        this.collectionNamesCacheOverflowErrorTracker = collectionNamesCacheOverflowErrorTracker;
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, String status, long elapsedTimeInNanoseconds) {
-        timerBuilder
```

Comment:
As this was written before, it was actually not thread-safe and tags from different threads could overwrite tags from other threads as `Timer.Builder.tag()` mutates the tags collection on the builder. Instead, here we reconstruct the instance on each command. I dug in a bit and it did not look expensive to recreate the builder. Let me know if that is not the case.